### PR TITLE
Polar limits enhancements

### LINKED ISSUE #328: thetamin/-max for polar plot
[Original report at SourceForge, opened Sun Jun 15 16:58:00 2008](https://sourceforge.net/tracker/?func=detail&aid=1994587&group_id=80706&atid=560723)

It would be nice if there is a possibility to display only a circular sector of a polar plot. 
At the moment there is always plotted the full circle (360 degree).

So I suggest a thetamax and thetamin parameter for the polar plot.

Thanks!

### LINKED ISSUE #1730: No grid for min values in polar plots
Grid ylines and ticks are not drawn for y <= 0 in polar plots

Ex :
no grid for min values in polar plots

The bug is true in matplotlib 1.1.1 and 1.2.0. Thanks.

from matplotlib.pyplot import figure, show
import numpy as np

x = np.arange(0,7,0.001)
y = np.arange(-2.0,5.0, 0.001)
fig = figure()

ax = fig.add_subplot(111, polar=True)
ax.plot(x, y)
ax.autoscale_view(True,True,True)

fig.show()
raw_input()

### LINKED ISSUE #4217: Feature request: offset radial origin for polar plots
I'd be really interested in having a nice way to produce polar plots where the origin of the radial axis is offset from the centre, something like this:

origin example

Having the radial offset is particularly useful for plotting polar histograms, which tend to suffer from the problem that long bars are over-emphasised relative to short bars because their areas being much larger due to the distorting effect of the polar transformation. Adding a radial offset helps to reduce the visual impact of this distortion.

There are of course hacky ways to achieve this sort of effect, e.g. offsetting the r coordinates of the data, then manually setting the radial ticks. I also came across [this SO question](http://stackoverflow.com/q/16605137/1461210). The accepted answer, using `floating_axes.GridHelperCurveLinear`, is close to what I'm looking for:

so_example

However, it's still a bit of a nasty hack, since it requires the radial ticks to be set manually.

How much work would it be to add a `set_r_offset()` method to allow one to control the relative radial position of the r-axis origin? I imagine it accepting a float between 0 and 1, where 0 would place the r origin exactly at the centre, and 1 would place it exactly at the edge of the plot.

### PR DESCRIPTION
**Pull request by QuLogic**
This is a ~~WIP~~ PR to add several enhancements to polar plots.
- [x] Allow negative radial grids and ticks. Negative radii are already plotted; they are simply shifted so that the minimum occurs at the centre. Fixes #1730 and #2203.
- [x] Allow offsetting the minimum radius location from the exact centre of the plot. This produces an annulus instead of a full circle. This is sort of done, but it's a bit of a hack. After reading up on transforms, I think this can maybe be worked out a little nicer. Fixes #4217.
- [x] Allow specifying minimum and maximum theta range to produce "wedges". Fixes #328.

### CODE REVIEW

**Comment by efiring** [comment]
The improvements you list are long overdue.  Thanks very much for taking this on.  I hope you will also be able to fix the zooming and panning, so that it works in display space rather than in data space, but preserves the aspect ratio.  In other words, it should be like using a magnifying glass to look at a map.

**Comment by QuLogic** [comment]
Where is zooming implemented? Obviously, it's not simple as just setting `can_zoom` to return `True`. Is it just a special case in `start/end_pan`?

**Comment by efiring** [comment]
It's confusing because there are two kinds of zoom: the zoom box (zoom-to-rect), and the continuous zoom/pan.  The latter is implemented in the PolarAxes.drag_pan() method, but for normal axes drag_pan and all the other zoom-to-rect and pan/zoom methods are in the backend_bases.py NavigationToolbar2.  There are PRs in the works to refactor some of these things.  @OceanWolf, can you provide some advice on how to proceed?  Now I'm thinking that maybe at least for the zoom-to-rect functionality, @QuLogic should defer action until the refactoring is merged.

**Comment by QuLogic** [comment]
Ah, looking at that code, it makes sense that polar axes don't zoom quite right (if I enable it). Instead of applying some sort of cropped zoom, it just modifies the x/y limits, which in polar axes does not exactly produce a box, and of course wouldn't work before when you couldn't modify the limits at all... Though once I get the plot to actually crop to the wedge instead of leaving the full circle, maybe that won't be so bad.

**Comment by QuLogic** [comment]
Well, I'm finally getting there...
output
You can use the zoom button to zoom in, though it produces wedges instead of a rectangular crop. The pan/zoom button hasn't been changed yet.

There are a few hacks though, which I'll have to figure out eventually. Right now, in order to update the background patch, I've had to make the `PolarAxes` pretend it's a parent to the view/radial limits. Maybe there's a simpler way around that. Also, you'll notice that the background patch does not seem to be clipping the data. Is there some kind of update that I need to call?

Theta text 1 and text 2 are swapped now, so the tests are going to break. I'm not sure if I can switch them back or whether their visibility should just be swapped.

The transforms were hacked on until they worked, so I'm pretty sure they can be optimized slightly.

**Comment by OceanWolf** [comment]
Don't have time to look at this right now, so very rough advice.  New toolbar has been delayed until next release due to review delays, but as I don't think this will go in either for then, not a problem.

Basically in terms of zoom/panning, just take a look at the code in master in ``backend_tools.py``, that contains everything you need to know about the new toolbar.  You can test it out on ``Gtk3`` and/or ``TkAgg`` by setting ``rcParam['toolbar'] = 'toolmanager'``

**Comment by QuLogic** [comment]
@OceanWolf: FYI, that does not work:
```python
  File ".../lib/matplotlib/pyplot.py", line 518, in figure
    **kwargs)
  File ".../lib/matplotlib/backends/backend_tkagg.py", line 84, in new_figure_manager
    return new_figure_manager_given_figure(num, figure)
  File ".../lib/matplotlib/backends/backend_tkagg.py", line 112, in new_figure_manager_given_figure
    figManager = FigureManagerTkAgg(canvas, num, window)
  File ".../lib/matplotlib/backends/backend_tkagg.py", line 545, in __init__
    backend_tools.add_tools_to_manager(self.toolmanager)
  File ".../lib/matplotlib/backend_tools.py", line 971, in add_tools_to_manager
    toolmanager.add_tool(name, tool)
  File ".../lib/matplotlib/backend_managers.py", line 221, in add_tool
    tool_cls = self._get_cls_to_instantiate(tool)
  File ".../lib/matplotlib/backend_managers.py", line 307, in _get_cls_to_instantiate
    globals(), locals(), [mod], -1)
ValueError: level must be >= 0
```

**Comment by QuLogic** [comment]
So this is nearly complete (WRT original goals) except for a couple issues:
 * When I update the axes patch, none of the data lines/points/etc. are clipped! How do I get the other patches to re-clip themselves?
 * The use of `_invalidate_internal` is a bit of a hack. It takes care of updating the axes patch dimensions, and the visibility of spines whenever any of the view parameters are changed. I was uncertain as to how the `viewLim` might be altered, so I found this to be the safest method. But if there is some sort of guarantee that it's only modified by method calls, I could make this more explicit.
 * No matter what I return for the horizontal or vertical alignments of the radial tick labels on the first axis, they always seem to be in the same place! The second axis appears to work nicely.

I will try and tackle the pan/zoom and add some change documentation as well, but it would be helpful to get some insight on the above first.

File: lib/matplotlib/projections/polar.py [hunk 1]
```diff
@@ -282,59 +432,99 @@ def _init_axis(self):
...
+        # The "real" view limit in consistent (theta, r) space, after applying
+        # all direction, offset, and limit options.
+        self._realViewLim = mtransforms.TransformedBbox(self._shiftedViewLim,
+                                                        self.transOffset)
+        self._realViewLim._parents[id(self)] = self
+
+        # Scale view limit into a bbox around the selected wedge. This may be
+        # smaller than the usual unit axes rectangle if not plotting the full
+        # circle.
+        self.axesLim = _WedgeBbox((0.5, 0.5), self._realViewLim)
```
Comment (QuLogic) [inline lib/matplotlib/projections/polar.py]: I'm not sure if `axesLim` is required to be unitized, but it isn't here, necessarily.

**Comment by QuLogic** [comment]
I added some "what's new" text.

@efiring I can probably implement the modified zoom/pan, but the API between backend and axis is not very nice. The trouble is that the backend seems to think that the only relevant information is the view limit. That makes no sense on a polar axis if you want to use cartesian zoom. If I implement pan/zoom using an additional transform (for example), then the home/back/forth buttons will not work.

**Comment by efiring** [comment]
Maybe the thing to do is skip the Cartesian zoom/pan for the present PR, and enter it as a new wish-list issue.  Then, if you are willing, you could attack it separately, making whatever larger changes are needed so that everything could work--presumably for other coordinate systems in addition to polar. @pelson, how have you handled zoom/pan in cartopy?  Any ideas about this?

**Comment by tacaswell** [comment]
I am a bit concerned by the number of test images that have been changed.

**Comment by QuLogic** [comment]
Because the `RadialLocator` no longer limits ticks to positive values, the test results now include a tick at 0.0.

**Comment by QuLogic** [comment]
BTW, I should probably note that those image changes were ignored by the default threshold. The only reason I noticed is because one test image was wrong, and the difference was ignored until the added 0.0 triggered a failure.

**Comment by efiring** [comment]
Those zero-ticks don't look good; maybe they need to be special-cased away.  I would consider them a regression.

**Comment by QuLogic** [comment]
I was debating that, but what should happen when you add an offset? Right now, you get a tick, but previously, you wouldn't.

**Comment by QuLogic** [comment]
> @pelson, how have you handled zoom/pan in cartopy? Any ideas about this?

IIUC, Cartopy "cheats" by always plotting the projected space, which is in Cartesian coordinates.

**Comment by efiring** [comment]
On 2015/07/22 9:07 AM, Elliott Sales de Andrade wrote:
> IIUC, Cartopy "cheats" by always plotting the projected space, which is
> in Cartesian coordinates.

That's what Basemap does, but I thought Cartopy was using the mpl 
transform machinery.  Maybe not.

**Comment by efiring** [comment]
@mdboom, do you have any thoughts about this?  The problem is that for non-affine projections like polar, we would like zoom and pan to work in Cartesian space.  Zooming radially inward in a polar plot is rarely useful, if ever.

**Comment by WeatherGod** [comment]
I should note that this issue is sort of similar to complaints about
mplot3d's "zooming" and "panning".

**Comment by QuLogic** [comment]
@efiring OK, the central tick is now not shown if you are plotting a full non-annular circle. In the other cases, I feel there is no benefit to hiding it. It looks like there are enough minor changes for GitHub to still think most of the test images changed, but if you look at the actual diff, they haven't really. I will probably have to squash the change to make it happy (and I will anyway to avoid the extraneous binary diffs.)

**Comment by tacaswell** [comment]
Can you avoid updating any of the test files unless they _must_ be updated?

**Comment by QuLogic** [comment]
> Can you avoid updating any of the test files unless they must be updated?

I can try that, but have you noticed that there are several (generally small) differences on master that are being ignored? polar_axes.png, polar_coords.png, polar_units_2.png, polar_wrap_*.png.

**Comment by QuLogic** [comment]
I squashed down so that only one existing test image has changed (and on purpose.) (x3 for different formats.)

**Comment by WeatherGod** [comment]
If need be, file a separate issue about a possible need for updating test
images. Let's keep this one focused on changes due to this PR.

**Comment by pelson** [comment]
> IIUC, Cartopy "cheats" by always plotting the projected space, which is in Cartesian coordinates.

Ultimately, at some point you have to use Cartesian coordinates if you want to put something on a monitor... :wink: 

> @pelson, how have you handled zoom/pan in cartopy?

The pan/zoom functionality is essentially handled for free by the fact that the whole transform stack is used to take you from geo-coordinates to axes space (and on to pixel space). The complexity with a non-rectangular patch though is that you have to deal with the background patch changing shape/extent - for that each time the patch is moved it is clipped to the axes' normalised [0, 1] space, and that subsequent clip path is used to clip any other artists on the axes. It is a testament to the mpl clip code that this is surprisingly quick and not too hard to implement.

**Comment by QuLogic** [comment]
> Ultimately, at some point you have to use Cartesian coordinates if you want to put something on a monitor... :wink:

Right, but the difference would be when that occurs. In Cartopy, the _axes_ are Cartesian and all the spines and such from the normal view are just hidden. Data must be plotted with a transform, or else it's plotted in Cartesian.

With the `PolarAxes` the data and the axes _are_ in polar coordinates, not Cartesian, so you don't need the transform. We could switch it to Cartopy's method, but I think that constitutes an API break.

**Comment by OceanWolf** [comment]
Do you have an example of how pan/zoom would work in polar?  I can't imagine how the selection works and how that pertains to the final result.

My instinct says to do both... with the new toolbar it should become very easy to add a toggle tool for example to toggle between global zoom/pan and local zoom/pan, where I use global to refer to the global/aka monitor coordinates, and local to refer to the local coordinate system, in this case polar... we can if needed, do this based on axis selection, i.e. if global != local then show toggle button, if not, hide it...

**Comment by efiring** [comment]
On 2015/07/25 2:04 AM, OceanWolf wrote:
> My instinct says to do both... with the new toolbar it should become
> very easy to add a toggle tool for example to toggle between global
> zoom/pan and local zoom/pan, where I use global to refer to the
> global/aka monitor coordinates, and local to refer to the local
> coordinate system, in this case polar

No need, too complicated.  At some point, just switch the behavior. 
What you are referring to as local simply makes no sense, period.  It 
was a misfeature.  I suspect it was put in because it seemed to make 
sense from a code point of view, not because there was a clear use case 
for it from any user's point of view.

I think the problem is that for general coordinate systems, doing 
zoom/pan reliably and well is not so easy.  I was testing cartopy's 
implementation on a polar map projection, and it didn't take long before 
it raised an exception.

**Comment by QuLogic** [comment]
OK, this is essentially done WRT the original goals. I am still looking at the Cartesian zoom, but I don't want to hold up this PR for it. The only issue is the clipping due to #4788.

**Comment by QuLogic** [comment]
So I wrote a [little demo](https://gist.github.com/QuLogic/bcefaafba3c8433e45f5); not sure if it should be added anywhere. I was able to see that `axesLim` really should be unitized, or else the title gets placed strangely.

There are some edge cases with grids when the origin radius is greater than the minimum radius, or if the theta range is greater than 360, but I'm not sure the tradeoff to get them working is worth it rather than just documenting them as Not Good Ideas.

**Comment by QuLogic** [comment]
Also, I don't know why the picking test fails; why would `CompositeGenericTransform` not have a `_parents` attribute since it ultimately derives from `TransformNode`?

**Comment by pelson** [comment]
> Also, I don't know why the pickling test fails; why would CompositeGenericTransform not have a _parents attribute since it ultimately derives from TransformNode?

I'm not certain. Maybe there is a circular transform reference which is causing the following line to fire too late? https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/transforms.py#L119

**Comment by QuLogic** [comment]
Unfortunately, I can't find any circular references, unless they occur outside of the `PolarAxes` class.

**Comment by mdboom** [comment]
The conversation in #4762 got me looking into this with respect to the clip path.  Is it intentional that the data extends beyond the axes?  It looks like the clipping is going still the original full circle of the polar plot rather than the wedge.

figure_1

Perhaps it's just not updating the clip path when it changes...

**Comment by mdboom** [comment]
I think the issue is that the wedge is updated in `draw`.  Does it have to be?  It seems like that could be done when the wedge properties change rather than on every draw.  However, you do have the comment "For backward compatibility" there, so there's quite likely something I'm missing.

Adding something like:

```
        for coll in (self.artists, self.collections, self.images, self.lines):
            for artist in coll:
                artist.set_clip_path(self.patch)
```

after the wedge changes shape in draw seems to address the issue.

**Comment by pbregener** [comment]
What's the status of this PR? The functionality implemented here would be greatly appreciated! :)

**Comment by QuLogic** [comment]
This won't be going in until 2.1, and there are a lot of test-related changes going into 2.0, so I've held out on updating until that's taken care of.

**Comment by tacaswell** [comment]
@QuLogic This needs another rebase.

**Comment by QuLogic** [comment]
@tacaswell waiting for #7873 also.

**Comment by QuLogic** [comment]
I split out #8944 from here, but I have this all rebased locally.

**Comment by QuLogic** [comment]
There are a lot of redundant test image changes here and it includes two other PR, but this is rebased now.

**Comment by QuLogic** [comment]
I don't really understand why this is not passing on CI; it works fine locally and I made sure to use the correct local FreeType setting.

**Comment by QuLogic** [comment]
It took entirely too long to figure out that when `reset_ticks` is called, I also needed to re-add the clip path on the ticks. Anyway, this seems to be back to working again.

**Comment by efiring** [comment]
@QuLogic Thank you for sticking with this.  These are wonderful improvements.

**Comment by QuLogic** [comment]
Thanks!

**Pull request merged**